cpp/include/cuspatial/experimental/detail/point_in_polygon.cuh

```diff
@@ -0,0 +1,214 @@
+
+#include <thrust/memory.h>
+
+#include <iterator>
+#include <type_traits>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @brief Kernel to test if a point is inside a polygon.
+ *
+ * Implemented based on Eric Haines's crossings-multiply algorithm:
+ * See "Crossings test" section of http://erich.realtimerendering.com/ptinpoly/
+ * The improvement in addenda is also addopted to remove divisions in this kernel.
+ *
+ * TODO: the ultimate goal of refactoring this as independent function is to remove
+ * src/utility/point_in_polygon.cuh and its usage in quadtree_point_in_polygon.cu. It isn't
+ * possible today without further work to refactor quadtree_point_in_polygon into header only
+ * API.
+ */
+template <class Cart2d,
+          class OffsetType,
+          class OffsetIterator,
+          class Cart2dIt,
+          class OffsetItDiffType = typename std::iterator_traits<OffsetIterator>::difference_type,
+          class Cart2dItDiffType = typename std::iterator_traits<Cart2dIt>::difference_type>
+__device__ inline bool is_point_in_polygon(Cart2d const& test_point,
+                                           OffsetType poly_begin,
+                                           OffsetType poly_end,
+                                           OffsetIterator ring_offsets_first,
+                                           OffsetItDiffType const& num_rings,
+                                           Cart2dIt poly_points_first,
+                                           Cart2dItDiffType const& num_poly_points)
+{
+  using T = iterator_vec_base_type<Cart2dIt>;
+
+  bool point_is_within = false;
+  // for each ring
+  for (auto ring_idx = poly_begin; ring_idx < poly_end; ring_idx++) {
+    int32_t ring_idx_next = ring_idx + 1;
+    int32_t ring_begin    = ring_offsets_first[ring_idx];
+    int32_t ring_end =
+      (ring_idx_next < num_rings) ? ring_offsets_first[ring_idx_next] : num_poly_points;
+
+    Cart2d b     = poly_points_first[ring_end - 1];
+    bool y0_flag = b.y > test_point.y;
+    bool y1_flag;
+    // for each line segment, including the segment between the last and first vertex
+    for (auto point_idx = ring_begin; point_idx < ring_end; point_idx++) {
+      Cart2d const a = poly_points_first[point_idx];
+      y1_flag        = a.y > test_point.y;
+      if (y1_flag != y0_flag) {
+        T run           = b.x - a.x;
+        T rise          = b.y - a.y;
+        T rise_to_point = test_point.y - a.y;
+
+        // The divergence here is introduced to avoid a potential division
+        // of `rise` on the rhs of the comparison operator. The assumption here
+        // is that the the division instruction costs more cycles than the
+        // divergence to synchronize.
+        if (rise > 0 && (test_point.x - a.x) * rise < run * rise_to_point)
+          point_is_within = not point_is_within;
+        else if (rise < 0 && (test_point.x - a.x) * rise > run * rise_to_point)
+          point_is_within = not point_is_within;
```

Comment:
Thinking about this more, I don't think there's any divergence here.  You can rewrite the code like this, which is probably closer to what the compiler would do (I hope).

```
auto lhs = test_point.x - a.x;
auto rhs = run * rise_to_point;
if ((rise > 0 and lhs < rhs) || (rise < 0 && lhs > rhs))
  point_is_within = not point_is_within;
```

And I think it might be clearer to use this code (assuming it is just as fast).